### ISSUE TITLE
io: Make the crate conform to the API guidelines

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3634]
# Rust API Guidelines Checklist

- **Naming** *(crate aligns with Rust naming conventions)*
  - ~[ ] Casing conforms to RFC 430 ([C-CASE])~
  - [x] Ad-hoc conversions follow `as_`, `to_`, `into_` conventions ([C-CONV])
  - [x] Getter names follow Rust convention ([C-GETTER])
  - [x] Methods on collections that produce iterators follow `iter`, `iter_mut`, `into_iter` ([C-ITER])
  - [x] Iterator type names match the methods that produce them ([C-ITER-TY])
  - [x] Feature names are free of placeholder words ([C-FEATURE])
  - [x] Names use a consistent word order ([C-WORD-ORDER])
- **Interoperability** *(crate interacts nicely with other library functionality)*
  - [x] Types eagerly implement common traits ([C-COMMON-TRAITS])
    - `Copy`, `Clone`, `Eq`, `PartialEq`, `Ord`, `PartialOrd`, `Hash`, `Debug`,
      `Display`, `Default`
  - [x] Conversions use the standard traits `From`, `AsRef`, `AsMut` ([C-CONV-TRAITS])
  - [x] Collections implement `FromIterator` and `Extend` ([C-COLLECT])
  - [x] Data structures implement Serde's `Serialize`, `Deserialize` ([C-SERDE])
  - [x] Types are `Send` and `Sync` where possible ([C-SEND-SYNC])
  - [ ] Error types are meaningful and well-behaved ([C-GOOD-ERR])
      - Still needs #3883
  - [x] Binary number types provide `Hex`, `Octal`, `Binary` formatting ([C-NUM-FMT])
  - [x] Generic reader/writer functions take `R: Read` and `W: Write` by value ([C-RW-VALUE])
- **Macros** *(crate presents well-behaved macros)*
- All these can be checked off once  #3872 is done
  - [ ] Input syntax is evocative of the output ([C-EVOCATIVE])
  - [ ] Macros compose well with attributes ([C-MACRO-ATTR])
  - [ ] Item macros work anywhere that items are allowed ([C-ANYWHERE])
  - [ ] Item macros support visibility specifiers ([C-MACRO-VIS])
  - [ ] Type fragments are flexible ([C-MACRO-TY])
- **Documentation** *(crate is abundantly documented)*
  - [ ] Crate level docs are thorough and include examples ([C-CRATE-DOC])
  - [ ] All items have a rustdoc example ([C-EXAMPLE])
  - [x] Examples use `?`, not `try!`, not `unwrap` ([C-QUESTION-MARK])
  - [x] Function docs include error, panic, and safety considerations ([C-FAILURE])
  - [x] Prose contains hyperlinks to relevant things ([C-LINK])
  - [x] Cargo.toml includes all common metadata ([C-METADATA])
    - authors, description, license, homepage, documentation, repository,
      keywords, categories
  - [x] Release notes document all significant changes ([C-RELNOTES])
  - [x] Rustdoc does not show unhelpful implementation details ([C-HIDDEN])
- **Predictability** *(crate enables legible code that acts how it looks)*
  - [x] Smart pointers do not add inherent methods ([C-SMART-PTR])
  - [x] Conversions live on the most specific type involved ([C-CONV-SPECIFIC])
  - [x] Functions with a clear receiver are methods ([C-METHOD])
  - [x] Functions do not take out-parameters ([C-NO-OUT])
  - [x] Operator overloads are unsurprising ([C-OVERLOAD])
  - [x] Only smart pointers implement `Deref` and `DerefMut` ([C-DEREF])
  - [x] Constructors are static, inherent methods ([C-CTOR])
- **Flexibility** *(crate supports diverse real-world use cases)*
  - [x] Functions expose intermediate results to avoid duplicate work ([C-INTERMEDIATE])
  - [x] Caller decides where to copy and place data ([C-CALLER-CONTROL])
  - [x] Functions minimize assumptions about parameters by using generics ([C-GENERIC])
  - [ ] Traits are object-safe if they may be useful as a trait object ([C-OBJECT])
      - tracked by #3833 
- **Type safety** *(crate leverages the type system effectively)*
  - [x] Newtypes provide static distinctions ([C-NEWTYPE])
  - [x] Arguments convey meaning through types, not `bool` or `Option` ([C-CUSTOM-TYPE])
  - [x] Types for a set of flags are `bitflags`, not enums ([C-BITFLAG])
  - [x] Builders enable construction of complex values ([C-BUILDER])
- **Dependability** *(crate is unlikely to do the wrong thing)*
  - [x] Functions validate their arguments ([C-VALIDATE])
  - [x] Destructors never fail ([C-DTOR-FAIL])
  - [x] Destructors that may block have alternatives ([C-DTOR-BLOCK])
- **Debuggability** *(crate is conducive to easy debugging)*
  - [x] All public types implement `Debug` ([C-DEBUG])
  - [x] `Debug` representation is never empty ([C-DEBUG-NONEMPTY])
- **Future proofing** *(crate is free to improve without breaking users' code)*
  - [x] Sealed traits protect against downstream implementations ([C-SEALED])
  - [x] Structs have private fields ([C-STRUCT-PRIVATE])
  - [x] Newtypes encapsulate implementation details ([C-NEWTYPE-HIDE])
  - [x] Data structures do not duplicate derived trait bounds ([C-STRUCT-BOUNDS])
- **Necessities** *(to whom they matter, they really matter)*
  - [x] Public dependencies of a stable crate are stable ([C-STABLE])
  - ~[ ] Crate and its dependencies have a permissive license ([C-PERMISSIVE])~

[C-CASE]: https://rust-lang.github.io/api-guidelines/naming.html#c-case
[C-CONV]: https://rust-lang.github.io/api-guidelines/naming.html#c-conv
[C-GETTER]: https://rust-lang.github.io/api-guidelines/naming.html#c-getter
[C-ITER]: https://rust-lang.github.io/api-guidelines/naming.html#c-iter
[C-ITER-TY]: https://rust-lang.github.io/api-guidelines/naming.html#c-iter-ty
[C-FEATURE]: https://rust-lang.github.io/api-guidelines/naming.html#c-feature
[C-WORD-ORDER]: https://rust-lang.github.io/api-guidelines/naming.html#c-word-order

[C-COMMON-TRAITS]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits
[C-CONV-TRAITS]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-conv-traits
[C-COLLECT]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-collect
[C-SERDE]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-serde
[C-SEND-SYNC]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-send-sync
[C-GOOD-ERR]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-good-err
[C-NUM-FMT]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-num-fmt
[C-RW-VALUE]: https://rust-lang.github.io/api-guidelines/interoperability.html#c-rw-value

[C-EVOCATIVE]: https://rust-lang.github.io/api-guidelines/macros.html#c-evocative
[C-MACRO-ATTR]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-attr
[C-ANYWHERE]: https://rust-lang.github.io/api-guidelines/macros.html#c-anywhere
[C-MACRO-VIS]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-vis
[C-MACRO-TY]: https://rust-lang.github.io/api-guidelines/macros.html#c-macro-ty

[C-CRATE-DOC]: https://rust-lang.github.io/api-guidelines/documentation.html#c-crate-doc
[C-EXAMPLE]: https://rust-lang.github.io/api-guidelines/documentation.html#c-example
[C-QUESTION-MARK]: https://rust-lang.github.io/api-guidelines/documentation.html#c-question-mark
[C-FAILURE]: https://rust-lang.github.io/api-guidelines/documentation.html#c-failure
[C-LINK]: https://rust-lang.github.io/api-guidelines/documentation.html#c-link
[C-METADATA]: https://rust-lang.github.io/api-guidelines/documentation.html#c-metadata
[C-HTML-ROOT]: https://rust-lang.github.io/api-guidelines/documentation.html#c-html-root
[C-RELNOTES]: https://rust-lang.github.io/api-guidelines/documentation.html#c-relnotes
[C-HIDDEN]: https://rust-lang.github.io/api-guidelines/documentation.html#c-hidden

[C-SMART-PTR]: https://rust-lang.github.io/api-guidelines/predictability.html#c-smart-ptr
[C-CONV-SPECIFIC]: https://rust-lang.github.io/api-guidelines/predictability.html#c-conv-specific
[C-METHOD]: https://rust-lang.github.io/api-guidelines/predictability.html#c-method
[C-NO-OUT]: https://rust-lang.github.io/api-guidelines/predictability.html#c-no-out
[C-OVERLOAD]: https://rust-lang.github.io/api-guidelines/predictability.html#c-overload
[C-DEREF]: https://rust-lang.github.io/api-guidelines/predictability.html#c-deref
[C-CTOR]: https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor

[C-INTERMEDIATE]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-intermediate
[C-CALLER-CONTROL]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control
[C-GENERIC]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-generic
[C-OBJECT]: https://rust-lang.github.io/api-guidelines/flexibility.html#c-object

[C-NEWTYPE]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-newtype
[C-CUSTOM-TYPE]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-custom-type
[C-BITFLAG]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-bitflag
[C-BUILDER]: https://rust-lang.github.io/api-guidelines/type-safety.html#c-builder

[C-VALIDATE]: https://rust-lang.github.io/api-guidelines/dependability.html#c-validate
[C-DTOR-FAIL]: https://rust-lang.github.io/api-guidelines/dependability.html#c-dtor-fail
[C-DTOR-BLOCK]: https://rust-lang.github.io/api-guidelines/dependability.html#c-dtor-block

[C-DEBUG]: https://rust-lang.github.io/api-guidelines/debuggability.html#c-debug
[C-DEBUG-NONEMPTY]: https://rust-lang.github.io/api-guidelines/debuggability.html#c-debug-nonempty

[C-SEALED]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed
[C-STRUCT-PRIVATE]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-private
[C-NEWTYPE-HIDE]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-newtype-hide
[C-STRUCT-BOUNDS]: https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-bounds

[C-STABLE]: https://rust-lang.github.io/api-guidelines/necessities.html#c-stable
[C-PERMISSIVE]: https://rust-lang.github.io/api-guidelines/necessities.html#c-permissive

**Comment by jamillambert** [comment]
Checked - Documentation examples ([C-QUESTION-MARK](https://rust-lang.github.io/api-guidelines/documentation.html#c-question-mark)). Can be checked off, there are no try! or unwrap.

**Comment by tcharding** [comment]
Scratched out items are checked off over here: https://github.com/rust-bitcoin/rust-bitcoin/issues/1649

**Comment by tcharding** [comment]
C-STABLE: Done. The only dependency is `internals` and grepping the API text files shows `internals` is not mistakenly in the public API.

**Comment by tcharding** [comment]
C-SEALED: Done. There are 4 traits. `IntoBoxDynDebug` is in a private `sealed` module. And the other 3 are meant to be implemented by users (`Read`, `BufRead`, `Write`)

**Comment by tcharding** [comment]
C-STRUCT-PRIVATE: Done. All structs have private fields except `Sink` which has no fields as expected.

**Comment by tcharding** [comment]
C-NEWTYPE-HIDE: Done. The only function parameters are generics, buffers (byte slices), or `position` (`u64`). No new types required by this lib.

**Comment by tcharding** [comment]
C-STRUCT-BOUNDS: Done. The only trait bounds are on `Take` and they look correct to me (and are not derivable traits).
```rust
pub struct Take<'a, R: Read + ?Sized> {
    reader: &'a mut R,
    remaining: u64,
}
```

**Comment by tcharding** [comment]
All checkmarks added to here.

**Comment by tcharding** [comment]
PR #3834 does:

- C-SEND-SYNC
- C-DEBUG
- C-DEBUG-NONEMPTY

**Comment by tcharding** [comment]
C-DTOR-FAIL and C-DTOR-BLOCK

There are no custom imlps of `Drop` so I believe both of these can be considered done.

**Comment by tcharding** [comment]
This crate does not contain regular data structures:

- `FromStd<T>`: Adapter
- `ToStd<T>`: Adapter
- `Cursor<T>`: Wrapper implements Read/Write and `pos()`
- `Sink`: Just to implement Read/Write
- `Take<'a, R: bitcoin_io::Read + ?core::marker::Sized>`: Adapter

This implies

- C-SERDE: Unnecessary, no `serde` support required by this crate

Note also:

- C-COMMON-TRAITS: Done in #3861
- C-GETTER: Done, used grep to confirm there are no functions with get in the name in units
- C-STABLE: Done, grepped the `api/io` directory for `internals`
- C-STRUCT-PRIVATE: Done, all non-error structs have private fields.

The following are based on https://github.com/rust-bitcoin/rust-bitcoin/issues/3632 and we just trusted that I got them right over there.

- C-DEREF: Done, checked with git grep Deref - no results.

- C-NO-OUT: Done. Checked with git grep '\&mut' | egrep -v 'mut fmt' | egrep -v 'mut self'
- C-FEATURE: Done, the only features are default, std, and alloc.

**Comment by tcharding** [comment]
All checkmarks added to here.

**Comment by tcharding** [comment]
A bunch of docs stuff checked while doing #3874 

Still to do: C-EXAMPLE and C-CRATE-DOC 

@apoelstra do you think we need to add anything more for the two above?

**Comment by apoelstra** [comment]
Aside from the macros which I complained about in #3872, I think we are ok for docs. For examples, we should definitely include an example of deserializing or serializing a Bitcoin object to a `std::io::Read`/`Write`, and of how to implement these traits properly on a new reader/writer (e.g. we could make a copy of io::Sink in the example called ByteCounter or something).

**Comment by tcharding** [comment]
Inspired #3878, thanks. I actually started this yesterday but the code I wrote today based on you're idea is way better.

**Comment by tcharding** [comment]
Done: C-CONV, manually checked all conversion functions.

**Comment by tcharding** [comment]
Done: C-ITER, there are no iterators in `io`.

**Comment by tcharding** [comment]
C-WORD-ORDER: Done, all types are correctly named

```
FromStd<T>
ToStd<T>
ErrorKind
Cursor<T>
Error
Sink
Take<'a, R: bitcoin_io::Read + ?core::marker::Sized
```

**Comment by tcharding** [comment]
Done:

- C-COLLECT: There are no collections in `io`.
- C-NUM-FMT: There are on binary number formats in `io`.
- C-RW-VALUE: There are no generic function parameters except in constructors.

**Comment by tcharding** [comment]
Done:

- C-NEWTYPE: The only function args are `Vec<u8>` and `u64` for `position` of a `Cursor`.

**Comment by tcharding** [comment]
C-VALIDATE: Done

- `read_exact`: `buf` is checked.
- `take`: `limit` is unchecked but excessive values are handled correctly (already tested implicitly in `read_to_limit` unit tests).
- `consume`: `amount` parameter documented to panic if too big
- Added some tests to `Take` (#3886)

**Comment by jamillambert** [comment]
Done:
 - C-SMART-PTR: No inherent methods on `Box<T>` and there is no ambiguity at the call site.
 - C-CONV-SPECIFIC: Checked all occurrences of `to_`, `from_`, `as_` and `into_`
 - C-METHOD: Used `gg -P '^pub fn .*?\('` and `gg -P '^pub const fn .*?\('`, none of the three functions should be methods on a type.
 - C-OVERLOAD: No implementations of operators with build-in syntax.
 - C-CTOR: The constructors are all static and inherent.

NB. The constructors all look like they conform to C-TOC, but it would be worth a double check, in particular the use of unsafe and Box.

**Comment by tcharding** [comment]
> C-SMART-PTR: No inherent methods on `Box<T>` and there is no ambiguity at the call site.

Just FTR I believe this one means 'if we have any smart pointers then they have no inherent methods' but we don't define any smart pointers in `io`. I hesitated at this one before because I was not totally sure of the difference between a smart pointer type and a wrapper type.

**Comment by tcharding** [comment]
Done:
- C-INTERMEDIATE: This crate does not have any functions that compute anything that may be useful to the caller. Check manually
- C-CALLER-CONTROL: No calls to `clone` in `io`
- C-GENERIC: The only function arguments are buffers and `position`, no need for more generics.

**Comment by tcharding** [comment]
Checked the claims made above by @jamillambert and marked as done.

**Comment by tcharding** [comment]
C-CONV-TRAITS: Done by checking docs against `std:io` docs for trait impls.

**Comment by tcharding** [comment]
If all the open PRs merge then just docs left to do - BOOM!

**Comment by tcharding** [comment]
No longer needed, `io` is off the list of soon-to-stabalise crates thanks to #3961
